### DISCUSSION TITLE
Explicit constraint terms, for complex quantified constraints

### DISCUSSION THREAD

**Discussion by Ericson2314** [ghc-proposals/ghc-proposals#467] [Ideas]
With quantified constraints, I am afraid there is need for abstract nonsense dictionaries that the solver has a hard time coming up with.

Here is an example from https://github.com/obsidiansystems/constraints-extras/pull/17 First keep in mind
```haskell
-- | Helper class to avoid impredicative type
class (forall c. ConstraintsForC f c => c a) => Extract f a
instance (forall c. ConstraintsForC f c => c a) => Extract f a
```
This function:
```haskell
weakenContra
  :: forall f g a
  .  (forall c. ConstraintsForC g c => ConstraintsForC f c)
  => Dict (Extract f a) -> Dict (Extract g a)
weakenContra Dict = Dict
```
doesn't work, as GHC gets confused. But
```haskell
weakenContra
  :: forall f g a
  .  (forall c. ConstraintsForC g c => ConstraintsForC f c)
  => Dict (Extract f a) -> Dict (Extract g a)
weakenContra @@x (Dict @@y) = Dict @@(a (\z => b y (x z))))
   where a :: Extract f a => (forall c. ConstraintsForC f c => c a)
         a = _ -- solver files in
         b :: (forall c. ConstraintsForC f c => c a) => Extract f a
         b = _
```
is more explicit and works just fine. In fact, we might even ditch the `Dict`s and write
```haskell
weakenContra
  :: forall f g a
  .  (forall c. ConstraintsForC g c => ConstraintsForC f c)
  => Extract f a => Extract g a -- note fat airrows
weakenContra x y = a (\z => b y (x z)))) 
   where a :: Extract f a => (forall c. ConstraintsForC f c => c a)
         a = _
         b :: (forall c. ConstraintsForC f c => c a) => Extract f a
         b = _
```
The obvious question is would this violate cannonicity, but I am pretty the answer is "no". There is no way to make a dictionary *from terms*, one can only write "combinators" that do lambda-calculus-y things to passed in dictionaries. I forget the exact semantic argument, but this means we are safe. We already don't complain if local constraints overlap with global instances because we know both construction give us equivalent instances --- this is the same sort of reasoning.

While it's a non-issue because we don't have any dictionary constructors/patterns anyways, one way to think about what were doing is an information flow security things where non-canonical things can depend on cannonical things, but not vice versa. This shows that "newtype" classes like `Extract` in fact could be given explicit constructors safely, since they only have canonical fields. But for simplicity, I don't take advantage of that for the second example and instead use `_` to just infer the wraps/unwraps I needed.

**Comment by goldfirere** [comment]
More generally, it seems you seek a way to offer GHC advice when it gets confused. I think this would be a great idea, but it would need quite a grand vision to imagine how to get the interactivity, etc., to work out.

**Reply by Ericson2314** [reply]
How more generally do you mean, exactly? I could imagine various IDEs making a nice interactive experience --- and that would be wonderful!, but it seems to core feature for GHC is a bit better bounded.

Beside helping GHC, one could argue in the vain of #448 this also provides a way to be "maximally explicit": letting the user write all the connectives and GHC just fills in the instances themselves is as explicit as we can be without risking canonicity violations.

**Comment by Ericson2314** [comment]
https://arxiv.org/abs/1807.11267 is about this
